src/runtime/chan.go

```diff
@@ -371,7 +497,61 @@ func chanClose(ch *channel) {
 //
 // TODO: do this in a round-robin fashion (as specified in the Go spec) instead
 // of picking the first one that can proceed.
-func chanSelect(recvbuf unsafe.Pointer, states []chanSelectState, blocking bool) (uintptr, bool) {
+func chanSelect(recvbuf unsafe.Pointer, states []chanSelectState) (uintptr, bool) {
+	if selected, ok := tryChanSelect(recvbuf, states); selected != ^uintptr(0) {
+		// one channel was immediately ready
+		return selected, ok
+	}
+
+	// construct blockchains
+	blockchains := make([]blockchain, len(states))
```

Comment:
That heap allocation is unfortunate. It would be great if this slice could be stack-allocated in the parent frame.